src/eddymotion/model/kernels.py

```diff
@@ -0,0 +1,174 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY kIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# We support and encourage derived works from this project, please read
+# about our expectations at
+#
+#     https://www.nipreps.org/community/licensing/
+#
+import numpy as np
+from sklearn.gaussian_process.kernels import Hyperparameter, Kernel
+
+
+class SphericalCovarianceKernel(Kernel):
+    """
+    Custom kernel based on spherical covariance function.
+
+    Parameters
+    ----------
+    lambda_s : float, default=1.0
+        Scale parameter for the covariance function.
+    a : float, default=1.0
+        Distance parameter where the covariance function goes to zero.
+    sigma_sq : float, default=1.0
+        Noise variance term.
+    lambda_s_bounds : pair of floats >= 0 or "fixed", default=(1e-5, 1e4)
+        The lower and upper bound on lambda_s.
+    a_bounds : pair of floats >= 0 or "fixed", default=(1e-5, np.pi)
+        The lower and upper bound on a.
+    sigma_sq_bounds : pair of floats >= 0 or "fixed", default=(1e-5, 1e4).
+    """
+
+    def __init__(
+        self,
+        lambda_s=2.0,
+        a=0.1,
+        sigma_sq=1.0,
+        lambda_s_bounds=(1e-5, 1e4),
+        a_bounds=(1e-5, np.pi),
+        sigma_sq_bounds=(1e-5, 1e4),
+    ):
+        self.lambda_s = lambda_s
+        self.a = a
+        self.sigma_sq = sigma_sq
+        self.lambda_s_bounds = lambda_s_bounds
+        self.a_bounds = a_bounds
+        self.sigma_sq_bounds = sigma_sq_bounds
+
+    @property
+    def hyperparameter_lambda_s(self):
+        return Hyperparameter("lambda_s", "numeric", self.lambda_s_bounds)
+
+    @property
+    def hyperparameter_a(self):
+        return Hyperparameter("a", "numeric", self.a_bounds)
+
+    @property
+    def hyperparameter_sigma_sq(self):
+        return Hyperparameter("sigma_sq", "numeric", self.sigma_sq_bounds)
+
+    def __call__(self, theta, theta_prime=None, eval_gradient=False):
+        """
+        Compute the kernel matrix.
+
+        Parameters
+        ----------
+        theta : array-like of shape (n_samples, n_samples)
+            Precomputed pairwise angles.
+        theta_prime : array-like of shape (n_samples, n_samples), default=None
+            Second input for the kernel function.
+        eval_gradient : bool, default=False
+            Determines whether the gradient with respect to the log of
+            the kernel hyperparameter is computed.
+
+        Returns
+        -------
+        K : array-like of shape (n_samples, n_samples)
+            Kernel matrix.
+        K_gradient : array-like of shape (n_samples, n_samples, n_dims), optional
+            The gradient of the kernel matrix with respect to the log of the
+            hyperparameters. Only returned when `eval_gradient` is True.
+        """
+        if theta_prime is not None:
+            theta = theta_prime
```

Comment:
This doesn't seem very useful - can we remove this `theta_prime` altogether?

Reply:
This is useful when making predictions. The predict method of the Gaussian Regressor of sklearn calls the kernel with self.X_train in addition to the data for which we are making predictions (see [line 439](https://github.com/scikit-learn/scikit-learn/blob/5491dc695/sklearn/gaussian_process/_gpr.py#L439)). The assumption that theta equals theta_prime holds only if we feed our predict method directly with the matrix containing the angles between the directions of the training set and the direction we want to predict. If we decide to feed our predict method with something else, we will need to adjust this and compute the angles within the kernel.

Reply:
By looking at the interface: https://github.com/scikit-learn/scikit-learn/blob/5bc86b885c2d4c9afde3b41ebbee46dcf8ca4254/sklearn/gaussian_process/kernels.py#L400

And the documentation there (for the compound kernel)

```
        X : array-like of shape (n_samples_X, n_features) or list of object, \
            default=None
            Left argument of the returned kernel k(X, Y)

        Y : array-like of shape (n_samples_X, n_features) or list of object, \
            default=None
            Right argument of the returned kernel k(X, Y). If None, k(X, X)
            is evaluated instead.
```

By reassigning this, you are  basically enforcing k(Y, Y) which I don't think makes sense.

Reply:
It makes sense when the kernel arguments are the matrices representing the differences between the bvec directions, not the bvec directions themselves. In this case, when making predictions, Y = θ(bvecs_test, bvecs_train). Therefore the kernel corresponds to k(x*, x) in equation 5 from Andersson 2015. However, I'm open to moving the computation of the angles inside the kernel and taking directly the bvec as arguments to make it more intuitive. Hence my comments [here](https://github.com/nipreps/eddymotion/pull/200#discussion_r1620664936) and [here](https://github.com/jhlegarreta/eddymotion/pull/2#discussion_r1613178501)

Reply:
> It makes sense when the kernel arguments are the matrices representing the differences between the bvec directions, not the bvec directions themselves. In this case, when making predictions, Y = θ(bvecs_test, bvecs_train). Therefore the kernel corresponds to k(x*, x) in equation 5 from [Andersson 2015](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4627362/).

Scikit learn is going to do k(x*, x) all by itself if Y=None -- the question is how you call kernel().

Now, if you want X to be only train samples, then something external to this PR will do just that (I would assume within the Gaussian Process PR).

I would be very surprised if this is the right place to do this.

Reply:
Instead of silently replacing the X input, we should do something explicit like this:

```suggestion
        if Y is not None:
            raise RuntimeError('Y should not be set')
```

Reply:
I may be missing something, but I do believe this https://github.com/nipreps/eddymotion/pull/203#discussion_r1631931390 is right.

`*x*` (and `*x'*`) are our bvecs, so I do not see where `Y` comes into play in the kernel. If we still disagree, maybe we can talk about it over a call and draw on a screen why we believe one thing of the other.

Reply:
>Now, if you want X to be only train samples, then something external to this PR will do just that (I would assume within the Gaussian Process PR) 
Yes, I assumed we would call the fit method of GaussianRegressor with the training set and the predict method with the test set using the remaining bvecs. When we do this, inside the [predict method](https://github.com/scikit-learn/scikit-learn/blob/5491dc695/sklearn/gaussian_process/_gpr.py#L439) of the GaussianRegressor, the kernel is called with two arguments: self.x_train and Y. If we directly input the angles difference, not the bvecs, Y is θ(bvecs_test, bvecs_train) (the angle difference between the bvec we want to predict and the bvecs from the training set. 

However, I strongly agree that it would be better to compute the angles inside the kernel. Additionally, if we input the angles difference directly into the predict method of sklearn, we will get a prediction for each angle. What we want, however, is a single prediction for each bvec. 

@jhlegarreta Sorry, it's a bit difficult to explain the problem with the dimensions in a github issue. I think organizing a meeting with a virtual whiteboard is a very good idea.